flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/catalog/CatalogCalciteSchema.java

```diff
@@ -0,0 +1,248 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.exceptions.DatabaseNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Schemas;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A mapping between Flink's catalog and Calcite's schema. This enables to look up and access tables
+ * in SQL queries without registering tables in advance. Databases are registered as sub-schemas in the schema.
+ * This mapping is modeled as a strict two-level reference structure for Flink in Calcite,
+ * the full path of tables and views is of format [catalog_name].[db_name].[meta-object_name].
+ */
+@Internal
+public class CatalogCalciteSchema implements Schema {
+	private static final Logger LOGGER = LoggerFactory.getLogger(CatalogCalciteSchema.class);
+
+	private final String catalogName;
+	private final ReadableCatalog catalog;
+
+	public CatalogCalciteSchema(String catalogName, ReadableCatalog catalog) {
+		this.catalogName = catalogName;
+		this.catalog = catalog;
+	}
+
+	/**
+	 * Looks up a sub-schema (database) by the given sub-schema name.
+	 *
+	 * @param schemaName Name of sub-schema to look up.
+	 * @return Sub-schema with a given dbName, or null.
+	 */
+	@Override
+	public Schema getSubSchema(String schemaName) {
+
+		if (catalog.databaseExists(schemaName)) {
+			return new DatabaseCalciteSchema(schemaName, catalog);
+		} else {
+			LOGGER.error(String.format("Schema %s does not exist in catalog %s", schemaName, catalogName));
+			throw new CatalogException(
+				new DatabaseNotExistException(catalogName, schemaName));
+		}
+	}
+
+	@Override
+	public Set<String> getSubSchemaNames() {
+		return new HashSet<>(catalog.listDatabases());
+	}
+
+	@Override
+	public Table getTable(String name) {
+		throw new UnsupportedOperationException(
+			"CatalogCalciteSchema does not support getTable()");
+	}
+
+	@Override
+	public Set<String> getTableNames() {
+		throw new UnsupportedOperationException(
```

Comment:
Can't we just return empty set?

Reply:
Since we will strictly follow the two structure as `<catalog>.<database>.<meta-object>`, does it make sense to allow trying to get tables and tables names out of the database layer? I personally see that as a bug if that happens in Flink planner.

What do you think?

Reply:
I agree that we should throw exception when `getTable` method is called. I see no reason though why can't we return empty table set here. I am not sure where calcite can call this method, we do not fully control the scope of this method. I think it doesn't harm us to return an `emptySet` here.

Reply:
ok, sounds good